### PR TITLE
Allow for special constant to be non-conformant with object's data type

### LINKED ISSUE #837: In text tables, validate attempts to match pattern associated with `data_type` before checking `Special_Constants`
### Checked for duplicates

Yes - I've already checked

May be part of #832 and/or related to #831, but I figured I'd open this for another example case if nothing else.

### 🐛 Describe the bug

When validating an ASCII table of times that uses "99:99:99" as a `Special_Constants/not_applicable_constant` value, validate fails due to the presence of the 99:99:99 value(s), with the message

> Value does not match its data type 'ASCII_Time': Could not parse 99:99:99 using these patterns 'hh:mm:ss[.ffffff][Z]'

### 🕵️ Expected behavior

I expected that instances of `Special_Constants` would get special treatment, and would be allowed as exceptions to `data_type` format rules.

### 📜 To Reproduce

1. Label an ASCII table that includes value(s) not normally allowed by the pattern associated with the applicable `<data_type>`
2. Include a definition of these value(s) as `<Special_Constants>` in the label
3. Run validate. Note resulting failure reason.

### 🖥 Environment Info

- Version of this software: v3.3.3 & v.3.5.0-SNAPSHOT
- Operating System: Ubuntu 20.04
- Java: OpenJDK 17.0.9

### 🩺 Test Data / Additional context

Sample data:
```
01:20:33
02:04:51
07:36:22
99:99:99
13:18:48
23:49:04
```

Relevant label section:
```xml
<Table_Delimited>
    <name>local_time</name>
    <offset unit="byte">0</offset>
    <parsing_standard_id>PDS DSV 1</parsing_standard_id>
    <records>6</records>
    <record_delimiter>Line-Feed</record_delimiter>
    <field_delimiter>Comma</field_delimiter>
    <Record_Delimited>
        <fields>1</fields>
        <groups>0</groups>
        <Field_Delimited>
            <name>time</name>
            <data_type>ASCII_Time</data_type>
            <maximum_field_length unit="byte">8</maximum_field_length>
            <description>Local (solar) time at boresight/target intercept</description>
            <Special_Constants>
                <not_applicable_constant>99:99:99</not_applicable_constant>
            </Special_Constants>
        </Field_Delimited>
    </Record_Delimited>
</Table_Delimited>
```

The sample data file, a functionally complete label for it, and a failing validate report can be found in [this attached gzipped tar file](https://github.com/NASA-PDS/validate/files/14389029/example_failure.tar.gz).

### 🦄 Related requirements

_No response_

### ⚙️ Engineering Details

Hopefully fairly simple fix to change ordering of checks here.

### PR DESCRIPTION
**Pull request by al-niessner**
## 🗒️ Summary
Added unit test then fix. Basically check on conversion if value matches a data type non-conforming special constant before logging error.

## ⚙️ Test Data and/or Report
See automated unit/regression testing below

## ♻️ Related Issues
Closes #837

### CODE REVIEW

File: src/main/java/gov/nasa/pds/tools/validate/SpecialConstantChecker.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        comparison = ((BigInteger) value)
+            .compareTo(SpecialConstantBitPatternTransforms.asBigInt(constants.getValidMinimum()));
+      } else {
+        if (constants.getValidMinimum().contains(".") || 
+            ((constants.getValidMinimum().contains("e") || constants.getValidMinimum().contains("E")) && 
+                !(constants.getValidMinimum().startsWith("0x") || constants.getValidMinimum().startsWith("0X") || constants.getValidMinimum().startsWith("16#")))) {
+          comparison = Double.valueOf(value.doubleValue())
+              .compareTo(SpecialConstantBitPatternTransforms.asBigDecimal(constants.getValidMinimum()).doubleValue());
+        } else {
+          Long con = SpecialConstantBitPatternTransforms.asBigInt(constants.getValidMinimum()).longValue();
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/tools/validate/SpecialConstantChecker.java:119]: "Long con" … reminds me of a past life 🤭

**Review by nutjob4life — APPROVED** [review]
👍

File: src/main/java/gov/nasa/pds/tools/validate/SpecialConstantChecker.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+public class SpecialConstantChecker {
+  /**
+   * Use this when the special constant may not conform to the type constraints
+   * meaning only string comparisons can be done.
+   * @param value
+   * @param constants
+   * @return
+   */
+  public static boolean isNonConformantSpecialConstant(String value, SpecialConstants constants) {
```
Comment (jordanpadams) [inline src/main/java/gov/nasa/pds/tools/validate/SpecialConstantChecker.java:17]: @al-niessner so per the [PDS4 IM Spec](https://pds.nasa.gov/datastandards/documents/im/v1/index_1L00.html), these are the special constants that do not require matching data types:
* error_constant - https://pds.nasa.gov/datastandards/documents/im/v1/index_1L00.html#attribute_pds_special_constants_pds_error_constant
* invalid_constant - https://pds.nasa.gov/datastandards/documents/im/v1/index_1L00.html#attribute_pds_special_constants_pds_invalid_constant
* missing_constant - https://pds.nasa.gov/datastandards/documents/im/v1/index_1L00.html#attribute_pds_special_constants_pds_missing_constant
* not_applicable_constant - https://pds.nasa.gov/datastandards/documents/im/v1/index_1L00.html#attribute_pds_special_constants_pds_not_applicable_constant
* saturated_constant - https://pds.nasa.gov/datastandards/documents/im/v1/index_1L00.html#attribute_pds_special_constants_pds_saturated_constant
* unknown_constant - https://pds.nasa.gov/datastandards/documents/im/v1/index_1L00.html#attribute_pds_special_constants_pds_unknown_constant

The rest all state that they must match the data type of the object and should throw an error if they do not match.

**Review by jordanpadams — CHANGES_REQUESTED** [review]

File: src/main/java/gov/nasa/pds/tools/validate/SpecialConstantChecker.java (hunk 2 above)
Comment (al-niessner) [reply src/main/java/gov/nasa/pds/tools/validate/SpecialConstantChecker.java:17]: done

**Comment by al-niessner** [comment]
@jordanpadams ready again

**Review by jordanpadams — APPROVED** [review]

**Comment by al-niessner** [comment]
@jordanpadams 

Your [merge of main](https://github.com/NASA-PDS/validate/pull/887/commits/e4621f25e5a50a10e2382c28935e98db21fa3fd2) into this branch broke everything.

I do not know how this happened but I suspect work was lost. Going to try fancy git to undo but whatever branch you merged into main then here will be lost. Hope you remember what it was. Worst case, all is lost on this branch and will have to be redone.

**Comment by al-niessner** [comment]
Deprecated this PR due to bad merge from main. See #893

**Pull request closed without merging**